### PR TITLE
RSDK-2214 remove pre map representation endpoints

### PR DESCRIPTION
**Pull request by JohnN193**
removes pre map representation endpoints. All C++ changes are the same as the [previously reverted commit](https://github.com/viamrobotics/viam-cartographer/pull/10), with the addition of changes for the go cartographer model.

ticket: https://viam.atlassian.net/browse/RSDK-2214

### CODE REVIEW

File: viam-cartographer_test.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
+// Package viamcartographer_test tests the functions that require injected components (such as robot and camera)
```
Comment (nicksanford) [inline viam-cartographer_test.go:1]: Why are we adding these tests when the ticket is about removing pre map representation endpoints?

File: viam-cartographer.go [hunk 2]
```diff
@@ -0,0 +1,447 @@
+// Package viamcartographer implements simultaneous localization and mapping.
+// This is an Experimental package.
+package viamcartographer
```
Comment (nicksanford) [inline viam-cartographer.go:3]: why is this being added?

**Review by nicksanford — CHANGES_REQUESTED** [review]

File: integration_test.go [hunk 3]
```diff
@@ -0,0 +1,387 @@
+// Package viamcartographer_test tests the functions that require injected components (such as robot and camera)
```
Comment (jeremyrhyde) [inline integration_test.go:1]: [minor] The comment here should be primarily focus on the type of tests this package runs, i.e. integration tests that test the full pipeline from providing fake data from an injected camera to a viam-cartographer binary and hitting the endpoints to confirm proper operation. Addition info such as using the internal package strikes me as superfluous.

File: internal/dim-2d/dim-2d.go [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+	lidar, err := lidar.New(deps, sensors, lidar2DIndex)
+	if err != nil {
+		return lidar, errors.Wrap(err, "configuring lidar camera error")
+	}
+
+	return lidar, nil
+}
+
+// ValidateGetAndSaveData makes sure that the provided sensor is actually a lidar and can
```
Comment (jeremyrhyde) [inline internal/dim-2d/dim-2d.go:55]: [q] Should the bring-up of a lidar be held in dim-2d as future work in cartographer to support 3D lidars would duplicate much of this. For example the `ValidateGetAndSaveData` would be identical as would the `GetAndSaveData`.

File: internal/dim-2d/dim-2d.go [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+	"go.viam.com/slam/sensors/lidar"
+	goutils "go.viam.com/utils"
+)
+
+const (
+	opTimeoutErrorMessage = "bad scan: OpTimeout"
+)
+
+// NewLidar returns a new lidar.Lidar for the 2D cartographer mode.
+func NewLidar(
```
Comment (jeremyrhyde) [inline internal/dim-2d/dim-2d.go:25]: [opt] `New2DLidar()`

File: internal/dim-2d/dim-2d.go [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+}
+
+// GetAndSaveData gets the pointcloud from the lidar and saves it to the provided data directory.
+// On success, it returns the absolute filepath where the data was saved, along with any error
+// associated with the data saving.
+func GetAndSaveData(ctx context.Context, dataDirectory string, lidar lidar.Lidar, logger golog.Logger) (string, error) {
+	ctx, span := trace.StartSpan(ctx, "viamcartographer::internal::dim2d::GetAndSaveData")
+	defer span.End()
+
+	pointcloud, err := lidar.GetData(ctx)
```
Comment (jeremyrhyde) [inline internal/dim-2d/dim-2d.go:108]: [q] Why are we calling `GetData` and not `NextPointCloud` anymore?

File: internal/dim-2d/dim-2d.go [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+	if err != nil {
+		return lidar, errors.Wrap(err, "configuring lidar camera error")
+	}
+
+	return lidar, nil
+}
+
+// ValidateGetAndSaveData makes sure that the provided sensor is actually a lidar and can
+// return pointclouds. It also ensures that saving the data to files works as intended.
+func ValidateGetAndSaveData(
```
Comment (jeremyrhyde) [inline internal/dim-2d/dim-2d.go:57]: Should validation be exposed or unexposed and called by of NewLidar

File: internal/dim-2d/dim-2d_test.go [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+
+		files, err := os.ReadDir(dataDir + "/data/")
+		test.That(t, len(files), test.ShouldEqual, 1)
+		test.That(t, err, test.ShouldBeNil)
+	})
+
+	testhelper.ClearDirectory(t, dataDir)
+}
+
+func TestValidateGetAndSaveData(t *testing.T) {
```
Comment (jeremyrhyde) [inline internal/dim-2d/dim-2d_test.go:90]: [minor] Can we add a test to check output if a bad lidar is given where say NextPointCloud returns an error as well as any other types of malformed lidars that were created back in out rdk unit tests?

**Comment by JohnN193** [comment]
PR went from 17 to 6 files changed, so I declare this actually safe to review now. sorry for the confusion!

File: go.mod [hunk 9]
```diff
@@ -12,11 +12,11 @@ require (
...
 	github.com/rhysd/actionlint v1.6.23
 	go.opencensus.io v0.24.0
-	go.viam.com/api v0.1.94
-	go.viam.com/rdk v0.2.28
+	go.viam.com/api v0.1.96
+	go.viam.com/rdk v0.2.31
 	go.viam.com/slam v0.1.33
 	go.viam.com/test v1.1.1-0.20220913152726-5da9916c08a2
-	go.viam.com/utils v0.1.17
+	go.viam.com/utils v0.1.18-0.20230327140716-bfeb34d89117
```
[the remark this answers is not in the document]
Comment (JohnN193) [reply go.mod:19]: The only thing I intentionally bumped was rdk, everything else was bumped as a dependency

**Review by kkufieta — COMMENTED** [review]
Did you test that cartographer still works via RDK & app.viam.com?

**Comment by JohnN193** [comment]
Yeah, all c++ code was tested on app. Unsure if I can test the go code at this current state

**Comment by kkufieta** [comment]
Yes, you can test the go code, and I recommend you do that since you're editing the go code. The steps are:
1. Register the `viam:slam:cartographer` model in RDK by editing services/slam/register/register.go like this:
  ```go
  // Package register registers all relevant slam models and also subtype specific functions
  package register
  
  import (
	  // for slam models.
	  "fmt"
  
	  viamcartographer "github.com/viamrobotics/viam-cartographer"
	  _ "go.viam.com/rdk/services/slam/fake"
  )
  
  func neverUsed() {
	  fmt.Println(viamcartographer.DefaultExecutableName)
  }
  ```
  (in go.mod, replace viam-cartographer with your local viam-cartographer repo path)

2. Change your config by replace `"model": "cartographer"` with `"model": "viam:slam:cartographer"`
3. Run & test cartographer

**Review by kkufieta — APPROVED** [review]
LGTM mod testing the go code via RDK & app.viam.com as described in my comment.

**Comment by JohnN193** [comment]
talked offline, I will not be testing the go code by changing my local branch of rdk to recognize it

**Pull request merged**